Set the default Render Context configuration per Platform

TomAtkinsonArm opened the pull request:
## Description

Currently all platforms running the samples have the same default render context configuration. API Samples force `VK_PRESENT_MODE_MAILBOX` and performance samples force `VK_PRESENT_MODE_FIFO`.

This PR proposes to set the default render context behaviour by what platform the application is being assumed that it is running on. Hence when the application runs on platforms presumed to be on desktop we prefer `VK_PRESENT_MODE_MAILBOX` to show discrepancy in framerate using different techniques and when using Android we prefer `VK_PRESENT_MODE_FIFO` which benefits are outlined in [this article](https://developer.samsung.com/sdp/blog/en-us/2019/07/26/vulkan-mobile-best-practice-how-to-configure-your-vulkan-swapchain)

Samples still have access to the `prepare_render_context` method where they can customize the context further. This change only targets the present mode default.

This PR helps solve part of #266 but does not include overriding the "vsync" through the CLI

## Notes

I have reintroduced the "Desktop Platform" under the justification that the code written in the later specialised platforms are already presumably running in a desktop environment. I understand that this hit a road bump before on the naming but I still see this as a useful place to remove duplication for these platforms.

Open to a discussion :)

## General Checklist:

Please ensure the following points are checked:

- [x] My code follows the [coding style](https://github.com/KhronosGroup/Vulkan-Samples/tree/master/CONTRIBUTING.md#Code-Style)
- [x] I have reviewed file [licenses](https://github.com/KhronosGroup/Vulkan-Samples/tree/master/CONTRIBUTING.md#Copyright-Notice-and-License-Template)
- [x] I have updated the header Copyright to reflect the current year (CI build will fail if Copyright is out of date)
- [x] I have commented any added functions (in line with Doxygen)
- [x] I have commented any code that could be hard to understand
- [x] My changes do not add any new compiler warnings
- [x] My changes do not add any new validation layer errors or warnings
- [x] I have used existing framework/helper functions where possible
- [x] My changes build on Windows, Linux, macOS and Android. Otherwise I have [documented any exceptions](https://github.com/KhronosGroup/Vulkan-Samples/tree/master/CONTRIBUTING.md#General-Requirements)
- [x] My changes do not add any regressions
- [x] I have tested every sample to ensure everything runs correctly
- [x] This PR describes the scope and expected impact of the changes I am making

## Sample Checklist (Does not apply)

If your PR contains a new or modified sample, these further checks must be carried out *in addition* to the General Checklist:
- [ ] I have tested the sample on at least one compliant Vulkan implementation
- [ ] If the sample is vendor-specific, I have [tagged it appropriately](https://github.com/KhronosGroup/Vulkan-Samples/tree/master/CONTRIBUTING.md#General-Requirements)
- [ ] I have stated on what implementation the sample has been tested so that others can test on different implementations and platforms
- [ ] Any dependent assets have been merged and published in downstream modules
- [ ] For new samples, I have added a paragraph with a summary to the appropriate chapter in the [samples readme](./../samples/README.md)

TomAtkinsonArm committed:
Add default render context

SaschaWillems reviewed (changes requested):
Sadly this does not work as intended on Windows for me. With this PR, all samples are now locked to V-Sync including those based on my framework.

It also seems to choose a different color space for my samples, so they now appear too bright.

Instancing in master:
![image](https://user-images.githubusercontent.com/10283127/119222149-31304b80-baf3-11eb-8f3e-3552463c8b86.png)

Instancing sample in this PR:
![image](https://user-images.githubusercontent.com/10283127/119222158-3ab9b380-baf3-11eb-80b2-4b0b8a541365.png)


The priority list for present modes on Desktop looks fine to me, but somehow it doesn't seem to actually use mailbox mode for on my RTX 2060.

gary-sweet commented on framework/CMakeLists.txt:
```diff
@@ -285,38 +291,38 @@ set(ANDROID_FILES
     # Source Files
-    platform/unix/unix_d2d_platform.cpp
-    platform/unix/direct_window.cpp) 
+    platform/desktop/unix/unix_d2d_platform.cpp
```
It looks as though you've categorised all unix platforms as desktop, which isn't the case.
Our STB platforms are embedded unix - very much not 'desktop'.

TomAtkinsonArm replied:
I had a think about this, it makes sense to leave them separate but I still wanted to remove the duplication. The platform now hosts some default functions and I had to edit the android platforms prepare call to remove the asynchronous step and have the same initialize -> prepare deterministic lifecycle for all platforms.

@SaschaWillems sorry this might need a reapproval

gary-sweet reviewed (changes requested):
I get a build error trying to build for embedded Linux D2D:

```
Vulkan-Samples/framework/platform/desktop/unix/direct_window.cpp:22:38: fatal error: platform/headless_window.h: No such file or directory
 #include "platform/headless_window.h"
```

TomAtkinsonArm commented:
> The priority list for present modes on Desktop looks fine to me, but somehow it doesn't seem to actually use mailbox mode for on my RTX 2060.

Looking deeper at how the render context configures the underlying swapchain the present mode was never set as a default. This meant that FIFO was always chosen... I fixed this by altering the default to the highest priority element in the priority lists

> It also seems to choose a different color space for my samples, so they now appear too bright

I missed a line you added which used a format that was not at the highest priority. I fixed this for API samples by overriding the preferred format. This now looks like your first image with my GPU

Please have another test with my latest changes :)

SaschaWillems commented:
LGTM. Being able to finally run esp. the performance samples without v-sync is great :)

gary-sweet commented on framework/CMakeLists.txt:
```diff
@@ -312,11 +311,11 @@ set(LINUX_D2D_FILES
-source_group("platform\\" FILES ${GLFW_FILES})
-source_group("platform\\windows" FILES ${WINDOWS_FILES})
+source_group("platform\\desktop\\" FILES ${GLFW_FILES})
+source_group("platform\\desktop\\windows" FILES ${WINDOWS_FILES})
```
These desktop bits look left over from the previous commit?

gary-sweet commented on framework/CMakeLists.txt:
```diff
@@ -312,11 +311,11 @@ set(LINUX_D2D_FILES
 source_group("platform\\android" FILES ${ANDROID_FILES})
-source_group("platform\\unix" FILES ${UNIX_FILES})
-source_group("platform\\unix" FILES ${LINUX_D2D_FILES})
+source_group("platform\\desktop\\unix" FILES ${UNIX_FILES})
```
Same here

gary-sweet commented on bldsys/cmake/template/entrypoint_main.cpp.in:
```diff
@@ -40,7 +40,7 @@ int main(int argc, char *argv[])
 	vkb::UnixD2DPlatform platform{argc, argv};
 #else
-#	include "platform/unix/unix_platform.h"
+#	include "platform/desktop/unix/unix_platform.h"
```
This "desktop" also needs removing I believe.

TomAtkinsonArm committed:
Remove redundant presentation support check

gary-sweet reviewed (changes requested):
I'm seeing the same thing that Sascha mentioned with the colour spaces being different now.

TomAtkinsonArm commented:
Hi, @gary-sweet I have attempted to reproduce the issue, wrong colour spaces, without success. Could you share the log when you run the sample please?

gary-sweet commented:
> Hi, @gary-sweet I have attempted to reproduce the issue, wrong colour spaces, without success. Could you share the log when you run the sample please?

@TomAtkinsonArm 

Log, as requested:
[info] Found GPU: VideoCore VI HW (V3D-620)
[warning] Couldn't find a discrete physical device, picking default GPU
[warning] Couldn't find a discrete physical device, picking default GPU
[info] Selected GPU: VideoCore VI HW (V3D-620)
[info] Dedicated Allocation enabled
[info] Performance query enabled
[info] Device supports the following requested extensions:
[info]          VK_KHR_get_memory_requirements2
[info]          VK_KHR_dedicated_allocation
[info]          VK_KHR_performance_query
[info]          VK_EXT_host_query_reset
[info]          VK_KHR_swapchain
[info] Surface supports the following surface formats:
[info]          VK_FORMAT_R8G8B8A8_UNORM, VK_COLORSPACE_SRGB_NONLINEAR_KHR
[info]          VK_FORMAT_A8B8G8R8_UNORM_PACK32, VK_COLORSPACE_SRGB_NONLINEAR_KHR
[info]          VK_FORMAT_R8G8B8A8_SRGB, VK_COLORSPACE_SRGB_NONLINEAR_KHR
[info]          VK_FORMAT_A8B8G8R8_SRGB_PACK32, VK_COLORSPACE_SRGB_NONLINEAR_KHR
[info] Surface supports the following present modes:
[info]          VK_PRESENT_MODE_FIFO_KHR
[info]          VK_PRESENT_MODE_MAILBOX_KHR
[warning] (Swapchain) Image extent (0, 0) not supported. Selecting (1280, 720).
[warning] (Swapchain) Surface format (VK_FORMAT_UNDEFINED, VK_COLORSPACE_SRGB_NONLINEAR_KHR) not supported. Selecting (VK_FORMAT_R8G8B8A8_SRGB, VK_COLORSPACE_SRGB_NONLINEAR_KHR).
[info] (Swapchain) Image usage flags: VK_IMAGE_USAGE_TRANSFER_SRC_BIT VK_IMAGE_USAGE_COLOR_ATTACHMENT_BIT 
[info] (Swapchain) Present mode selected: VK_PRESENT_MODE_MAILBOX_KHR
[warning] (Swapchain) Surface format (VK_FORMAT_B8G8R8A8_UNORM, VK_COLORSPACE_SRGB_NONLINEAR_KHR) not supported. Selecting (VK_FORMAT_R8G8B8A8_SRGB, VK_COLORSPACE_SRGB_NONLINEAR_KHR).
[info] Depth format selected: VK_FORMAT_D32_SFLOAT
[info] Depth format selected: VK_FORMAT_D32_SFLOAT
[info] Depth format selected: VK_FORMAT_D32_SFLOAT
[info] Depth format selected: VK_FORMAT_D32_SFLOAT

It looks like you have hard coded BGRA, where before it was also looking for RGBA.

TomAtkinsonArm commented:
Ah it may be that your device does not support the BGRA surface format?

I chose BGRA as this is the [requested format in the ApiVulkanSample class](https://github.com/KhronosGroup/Vulkan-Samples/blob/2757f4fcf78b2f1659425f58a4e33c18973b8d09/framework/api_vulkan_sample.cpp#L94)

This means that the framework is falling back to the first supported format in the priority list which is RGBA in this case? Which would explain why you are left with the second image in Sascha's examples... It doesn't explain how it was displayed correctly before the change though...

gary-sweet commented:
> Ah it may be that your device does not support the BGRA surface format?
> 
> I chose BGRA as this is the [requested format in the ApiVulkanSample class](https://github.com/KhronosGroup/Vulkan-Samples/blob/2757f4fcf78b2f1659425f58a4e33c18973b8d09/framework/api_vulkan_sample.cpp#L94)
> 
> This means that the framework is falling back to the first supported format in the priority list which is RGBA in this case? Which would explain why you are left with the second image in Sascha's examples... It doesn't explain how it was displayed correctly before the change though...

Hmm. The log looks exactly the same on the master branch without this change, so maybe this is a red-herring.

gary-sweet commented:
It does look like you've swapped the priority of UNORM and SRGB formats though.

gary-sweet commented on framework/platform/platform.cpp:
```diff
@@ -153,6 +154,28 @@ void Platform::run()
+	auto context = std::make_unique<RenderContext>(device, surface, window->get_width(), window->get_height());
+
+	context->set_surface_format_priority({{VK_FORMAT_R8G8B8A8_SRGB, VK_COLOR_SPACE_SRGB_NONLINEAR_KHR},
+	                                      {VK_FORMAT_B8G8R8A8_SRGB, VK_COLOR_SPACE_SRGB_NONLINEAR_KHR},
```
These are in a different order to the original code.

TomAtkinsonArm replied:
The surface priority was moved for convenience with the presentation modes. I missed the fact that one prioritises UNORM and the other sRGB. I'll fix this later today :)

gary-sweet commented on framework/platform/platform.cpp:
```diff
@@ -153,6 +154,28 @@ void Platform::run()
+	                                      {VK_FORMAT_R8G8B8A8_UNORM, VK_COLOR_SPACE_SRGB_NONLINEAR_KHR},
+	                                      {VK_FORMAT_B8G8R8A8_UNORM, VK_COLOR_SPACE_SRGB_NONLINEAR_KHR}});
+
+	context->request_image_format(VK_FORMAT_R8G8B8A8_SRGB);
```
And this is now sRGB rather than UNORM.

TomAtkinsonArm commented on framework/vulkan_sample.cpp:
```diff
@@ -114,14 +114,7 @@ bool VulkanSample::prepare(Platform &platform)
-	render_context->set_surface_format_priority({{VK_FORMAT_R8G8B8A8_SRGB, VK_COLOR_SPACE_SRGB_NONLINEAR_KHR},
-	                                             {VK_FORMAT_B8G8R8A8_SRGB, VK_COLOR_SPACE_SRGB_NONLINEAR_KHR},
-	                                             {VK_FORMAT_R8G8B8A8_UNORM, VK_COLOR_SPACE_SRGB_NONLINEAR_KHR},
-	                                             {VK_FORMAT_B8G8R8A8_UNORM, VK_COLOR_SPACE_SRGB_NONLINEAR_KHR}});
```
Ah you are right, VulkanSample and ApiVulkanSample request formats differently

TomAtkinsonArm commented:
Hey @SaschaWillems, VK_FORMAT_B8G8R8A8_UNORM is requested but VK_FORMAT_R8G8B8A8_UNORM is higher in the priority list? With my latest change would this be better expressed by moving VK_FORMAT_B8G8R8A8_UNORM higher in the list?

[The original function which I am referencing](https://github.com/KhronosGroup/Vulkan-Samples/blob/2757f4fcf78b2f1659425f58a4e33c18973b8d09/framework/api_vulkan_sample.cpp#L83)

TomAtkinsonArm committed:
Allow custom creation of the render context

TomAtkinsonArm commented:
@gary-sweet the surface format lists should now be correct for API and framework samples, see `create_render_context()`

gary-sweet reviewed (approved):
Thanks. Looks much better now.

gpx1000 reviewed (approved):
Works here
